### DISCUSSION TITLE
Pinia/Vue3 -- does this usage look okay? any negative consequences?

### DISCUSSION THREAD

**Discussion by iambumblehead** [urql-graphql/urql#2587] [Q&A] [answered]
The first question: **is the sample below an acceptable way to setup an urql-using Pinia store?** Here the data would be used in the 'setup' method of a component that renders user info and maybe other details added to the query later, 
```javascript
import { ref } from 'vue';
import { defineStore } from 'pinia';
import { useQuery } from '@urql/vue';

export default defineStore( 'user', () => {
  const user = ref( null );

  ( async () => {
    const userRes = await useQuery({
      query: `{
        getSelf {
          id,
          name,
          email
        }
      }`
    });

    user.value = userResData.data && userResData.data.value.getSelf;
  })();

  return {
    user
  };
});
```
The second question is, **is there an interface or a convenient way of accessing deeply nested values without using "value" lookups?** There are situations where it would be nice to parse deeply nested data programmatically and I'm not sure how to do this with the type of data returned to me by urql
```javascript
[ usersActive, usersInactive, usersDeleted ]
  = urqlRes.data.game.skatepark.tournament.usersAll.reduce( separateUsersByStatus );
```

**Comment by PatrykWalach** [comment, marked as answer]
If this generates any errors `useQuery` probably uses some hooks unsupported inside pinia store, for example: `onMounted`.
I'd also expect the hook not to work outside of vue components since it depends on `injecting` the client.

However if this hook works inside `pinia` I wouldn't use await
```ts
export default defineStore('user', () => {
  const { data } = await useQuery({
    query: `{
      getSelf {
        id,
        name,
        email
      }
    }`
  })

  const user = computed(() => data.value.getSelf)

  return {
    user
  }
})
```
If the hook doesn't work you can just [subscribe ](https://formidable.com/open-source/urql/docs/api/core/#clientquery) to the query 
```ts
import { pipe, subscribe } from 'wonka';


export default defineStore('user', ({ client }) => { // you probably want to inject client here to simplify testing
  const user = ref(null)

  pipe(
    client.query(`{
      getSelf {
        id,
        name,
        email
      }
    }`),
    subscribe(({ data }) => {
      user.value = data && data.getSelf
    })
  )

  return {
    user
  }
})
```
